File: packages/dotnet-api/Source/Features/RuntimeEvents/Models/RuntimeEventTypes.cs
namespace Source.Features.RuntimeEvents.Models;

/// <summary>
/// String constants for the <see cref="RuntimeEvent.Type"/> column. Exposed as
/// <c>const string</c> (not an enum) so:
///
/// <list type="bullet">
///   <item>The daemon and backend can both emit the same wire values without
///         a shared C# enum — the daemon is a separate process / repo and
///         consumes these as JSON strings.</item>
///   <item>Adding a new event type is one line; no migration, no value-shift
///         risk on existing rows.</item>
///   <item>Switch statements and pattern matching get exhaustive-style ergonomics
///         (constant case labels) without forcing every consumer to import an
///         enum.</item>
/// </list>
///
/// <para>Taxonomy mirrors the runtime-spec-v2 specification's "Event taxonomy"
/// section. Keep this list in sync with the daemon's emitter constants — the
/// spec is the source of truth, this file is a mirror.</para>
/// </summary>
public static class RuntimeEventTypes
{
    // -------- Bootstrap stages --------
    public const string BootstrapStageStarted = "BootstrapStageStarted";
    public const string BootstrapStageCompleted = "BootstrapStageCompleted";
    public const string BootstrapStageFailed = "BootstrapStageFailed";

    // -------- Install (apt / mise / curl-pipe-bash snippets) --------
    public const string InstallStarted = "InstallStarted";
    public const string InstallCompleted = "InstallCompleted";
    public const string InstallSkipped = "InstallSkipped";
    public const string InstallFailed = "InstallFailed";

    // -------- Per-boot setup commands (npm install, migrations, …) --------
    public const string SetupCommandStarted = "SetupCommandStarted";
    public const string SetupCommandCompleted = "SetupCommandCompleted";
    public const string SetupCommandFailed = "SetupCommandFailed";

    /// <summary>
    /// Batched stdout/stderr lines from install or setup bash, emitted on a
    /// ~2-second flush cadence (or whenever the per-stream buffer reaches 50
    /// lines, whichever comes first). Lets the Timeline tab show live progress
    /// during long-running steps like <c>dotnet restore</c> or
    /// <c>mise install</c> instead of going silent between the
    /// <see cref="InstallStarted"/> / <see cref="SetupCommandStarted"/>
    /// bookends and their matching <c>Completed</c> events. Persisted via the
    /// usual <c>RuntimeEvents</c> pipeline so the operator can scroll back
    /// through what happened after the fact. Payload:
    /// <c>{ stage: "install"|"setup", stream: "stdout"|"stderr",
    /// lines: string[], batchedAt: ISO8601, lineCount: int,
    /// flushReason: "interval"|"size" }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Info"/>.
    /// </summary>
    public const string BootstrapOutputChunk = "BootstrapOutputChunk";

    // -------- Supervised service lifecycle --------
    public const string ServiceStarting = "ServiceStarting";
    public const string ServiceRunning = "ServiceRunning";
    public const string ServiceCrashed = "ServiceCrashed";
    public const string ServiceRestarted = "ServiceRestarted";

    /// <summary>
    /// Pre-flight guard tripped: a service declared one or more
    /// <c>requiredEnv</c> keys (from its <c>ServiceSpec.RequiredEnv</c>) that
    /// are absent or empty in the runtime's env snapshot at start time. The
    /// daemon's <c>StartingServicesStage</c> SKIPS supervisord registration for
    /// that service rather than letting it boot and crash-loop on the missing
    /// variable, then emits this event. The runtime still reaches Online
    /// (degraded) so the operator can add the secret in the Environment tab and
    /// restart the affected service. Payload:
    /// <c>{ serviceName, missingEnvVars: string[] }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/>.
    /// </summary>
    public const string ServiceEnvMissing = "ServiceEnvMissing";

    /// <summary>
    /// Per-service active healthcheck command (the <c>healthcheck.command</c>
    /// declared on a <c>ServiceSpec</c>) returned exit 0 within the stage's
    /// deadline. Emitted by the daemon's <c>StartingServicesStage</c> the first
    /// time a probe succeeds; the inner probe loop then breaks for that
    /// service. Payload: <c>{ serviceName, durationMs, lastExitCode }</c>
    /// where <c>durationMs</c> is measured from when the probe loop started
    /// and <c>lastExitCode</c> is always 0 (kept for symmetry with the
    /// <see cref="ServiceHealthcheckTimedOut"/> payload). Severity:
    /// <see cref="RuntimeEventSeverity.Info"/>.
    /// </summary>
    public const string ServiceHealthy = "ServiceHealthy";

    /// <summary>
    /// Per-service active healthcheck command never returned exit 0 within
    /// the stage's deadline. Emitted by the daemon's
    /// <c>StartingServicesStage</c>; the stage does NOT fail when this is
    /// emitted — the healthcheck is advisory once supervisord reports the
    /// service RUNNING. Payload: <c>{ serviceName, deadlineMs, lastExitCode,
    /// lastStdoutTail, lastStderrTail }</c> where the tails are capped at
    /// ~2KB each (curl / pg_isready outputs are short by nature). Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/> — process-alive is still
    /// required, but the post-RUNNING probe is best-effort.
    /// </summary>
    public const string ServiceHealthcheckTimedOut = "ServiceHealthcheckTimedOut";

    /// <summary>
    /// Per-service active healthcheck command returned non-zero for a single
    /// probe iteration. Bucketed emission: only emitted when the exit code
    /// transitions from the previous probe OR every 5th attempt — without
    /// that throttle a 180s deadline at a 5s interval would emit 36 events
    /// per service. Payload:
    /// <c>{ serviceName, attemptCount, exitCode, stdoutTail, stderrTail,
    /// attemptedAt }</c>. Severity: <see cref="RuntimeEventSeverity.Info"/>
    /// (debug-grade signal; useful for diagnosing flaky probes but not an
    /// operator-actionable alert on its own).
    /// </summary>
    public const string ServiceHealthcheckProbeFailed = "ServiceHealthcheckProbeFailed";

    /// <summary>
    /// Batched stdout/stderr lines from a supervised service captured during
    /// the <c>starting-services</c> bootstrap window (between supervisord
    /// <c>addService</c> returning and the stage emitting its final result).
    /// Same flush cadence as <see cref="BootstrapOutputChunk"/> (~2s or
    /// 50 lines, whichever comes first), but scoped to one service so the
    /// Timeline can label the lines with the source program. Payload:
    /// <c>{ serviceName, stream: "stdout"|"stderr", lines: string[],
    /// batchedAt: ISO8601, lineCount: int, flushReason: "interval"|"size" }</c>.
    /// Severity: <see cref="RuntimeEventSeverity.Info"/>. Tailing stops when
    /// the stage returns — post-Online live-tail is a separate concern handled
    /// by the service-logs SignalR stream.
    /// </summary>
    public const string ServiceOutputChunk = "ServiceOutputChunk";

    // -------- Spec delta application --------
    public const string SpecDeltaApplied = "SpecDeltaApplied";
    public const string SpecDeltaFailed = "SpecDeltaFailed";
    public const string SpecValidationFailed = "SpecValidationFailed";

    /// <summary>
    /// Degraded-online bootstrap (spec <c>self-healing-runtime-specs</c>, card
    /// D1). A NON-CRITICAL (spec) bootstrap stage — Install, RunningSetup, or
    /// StartingServices — failed deterministically, so the daemon's
    /// <c>BootstrapOrchestrator</c> recorded a <c>BootIssue</c> and CONTINUED
    /// instead of aborting the boot. One event is emitted per boot issue just
    /// before <c>ReportReadyStage</c> runs, so the runtime still reaches Online
    /// but in a <c>Degraded</c> <c>SpecHealth</c> state. The amber degraded
    /// banner on the runtime UI and the "let agent fix it" self-heal loop key
    /// off these events. Payload:
    /// <c>{ stage, service?, reason, detail?, occurredAt }</c> (mirrors the
    /// daemon's <c>BootIssue</c> shape). Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/> — the runtime is alive, the
    /// spec just didn't fully apply.
    /// </summary>
    public const string SpecDegraded = "SpecDegraded";

    // -------- Observability foundation (super-admin polish) --------
    //
    // These constants are added by the observability-foundation card so swagger
    // surfaces the type names that daemon + lifecycle-job cards will emit. Each
    // is mirrored in the daemon's emitter constants (or backend Hangfire job)
    // by the dependent cards — this file is the single source of truth for the
    // wire string. See spec `runtime-observability-super-admin` Section B for
    // payload shapes and emission rules.

    /// <summary>
    /// Service supervised by supervisord declared FATAL (exhausted
    /// <c>startretries</c>) without ever reaching RUNNING. Emitted by the
    /// daemon's <c>ServiceStatusPoller</c>. Payload:
    /// <c>{ serviceName, attemptCount, finalState, exitStatus, spawnErr,
    /// stderrTailLines }</c>. Severity: <see cref="RuntimeEventSeverity.Error"/>.
    /// </summary>
    public const string ServiceFailedToStart = "ServiceFailedToStart";

    /// <summary>
    /// Daemon attempted to ack a <c>RuntimeSpecDeltaApplied</c> result but the
    /// SignalR send rejected (typically payload-too-large or transport blow-up).
    /// Emitted by the daemon's <c>RuntimeSpecApplier.#sendAck</c> error path.
    /// Payload: <c>{ proposalId, errorMessage, ackPayloadBytes,
    /// signalrConnectionState }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/>.
    /// </summary>
    public const string SpecApplyAckFailed = "SpecApplyAckFailed";

    /// <summary>
    /// Our DB's view of a runtime's state disagrees with what Fly's machines
    /// API returns. Emitted by the backend's <c>FlyDriftPollerJob</c> on each
    /// 60-second tick where the live machine state doesn't match our stored
    /// runtime state. Payload: <c>{ flyState, ourState, machineId, region }</c>.
    /// Severity: <see cref="RuntimeEventSeverity.Warning"/>.
    /// </summary>
    public const string RuntimeFlyDriftDetected = "RuntimeFlyDriftDetected";

    /// <summary>
    /// Cloudflared tunnel for this runtime stopped responding (health check
    /// HEAD against the tunnel hostname failed). Emitted by the daemon's
    /// <c>CloudflaredController</c> failure-transition path. Payload:
    /// <c>{ hostname, lastError, lastSuccessAt }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/>.
    /// </summary>
    public const string CloudflaredTunnelDown = "CloudflaredTunnelDown";

    /// <summary>
    /// Cloudflared tunnel for this runtime is back up after a prior
    /// <see cref="CloudflaredTunnelDown"/>. Emitted by the daemon's
    /// <c>CloudflaredController</c> recovery-transition path. Payload:
    /// <c>{ hostname, lastError, lastSuccessAt }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Info"/>.
    /// </summary>
    public const string CloudflaredTunnelUp = "CloudflaredTunnelUp";

    /// <summary>
    /// Backend's <c>RespawnRuntimeJob</c> decided to respawn a crashed runtime
    /// (destroy old Fly machine + create replacement). Emitted before the new
    /// boot is triggered. Payload: <c>{ attemptNumber, lastFailureReason,
    /// lastFailureMessage, secondsSinceLastHeartbeat }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Warning"/>.
    /// </summary>
    public const string RuntimeRespawnTriggered = "RuntimeRespawnTriggered";

    /// <summary>
    /// Disk pressure on the runtime's <c>/data</c> volume transitioned to
    /// critical. Emitted by the daemon's <c>DiskMonitor</c> on the
    /// non-critical → critical edge. Payload:
    /// <c>{ usedBytes, totalBytes, level, mountpoint }</c>. Severity:
    /// <see cref="RuntimeEventSeverity.Error"/>.
    /// </summary>
    public const string DiskPressureCritical = "DiskPressureCritical";
}

File: packages/daemon/src/events/RuntimeEventTypes.ts
// RuntimeEventTypes — TypeScript mirror of the C# constants in
// `/workspace/packages/dotnet-api/Source/Features/RuntimeEvents/Models/RuntimeEventTypes.cs`.
//
// Wire values are plain strings (not an enum) so the daemon and backend can
// evolve the taxonomy without a coordinated migration: adding a new constant
// is one line on each side, dropping one is a deprecation + grep on each side.
//
// Source of truth is the "Event taxonomy" section of runtime-spec-v2.md. Keep
// these in lockstep with the C# RuntimeEventTypes class — drift will silently
// degrade the Timeline tab in the runtime drawer (the backend filter/query
// columns key off these exact strings).

export const RuntimeEventTypes = {
  // -------- Bootstrap stages --------
  BootstrapStageStarted: 'BootstrapStageStarted',
  BootstrapStageCompleted: 'BootstrapStageCompleted',
  BootstrapStageFailed: 'BootstrapStageFailed',

  // -------- Install (apt / mise / curl-pipe-bash snippets) --------
  InstallStarted: 'InstallStarted',
  InstallCompleted: 'InstallCompleted',
  InstallSkipped: 'InstallSkipped',
  InstallFailed: 'InstallFailed',

  // -------- Per-boot setup commands (npm install, migrations, …) --------
  SetupCommandStarted: 'SetupCommandStarted',
  SetupCommandCompleted: 'SetupCommandCompleted',
  SetupCommandFailed: 'SetupCommandFailed',

  /**
   * Batched stdout/stderr lines from install or setup bash. The daemon's
   * `BootstrapOutputBatcher` flushes ~every 2 seconds (or whenever the
   * per-stream buffer reaches 50 lines) and emits one event per stream per
   * flush. Lets the Timeline tab show live progress during long-running steps
   * like `dotnet restore` or `mise install`. Mirrors
   * `Source.Features.RuntimeEvents.Models.RuntimeEventTypes.BootstrapOutputChunk`
   * on the backend.
   */
  BootstrapOutputChunk: 'BootstrapOutputChunk',

  // -------- Supervised service lifecycle --------
  ServiceStarting: 'ServiceStarting',
  ServiceRunning: 'ServiceRunning',
  ServiceCrashed: 'ServiceCrashed',
  ServiceRestarted: 'ServiceRestarted',

  /**
   * Pre-flight guard: a service declared one or more `requiredEnv` keys that
   * are absent (or empty) in the runtime's env snapshot at start time. The
   * daemon SKIPS supervisord registration for that service rather than letting
   * it boot and crash-loop on the missing var. Emitted by
   * `StartingServicesStage` before `addService`. Payload:
   * `{ serviceName, missingEnvVars: string[] }`. Severity: Warn (the runtime
   * still comes Online — degraded — so the operator can add the secret and
   * restart the service). Mirrors
   * `Source.Features.RuntimeEvents.Models.RuntimeEventTypes.ServiceEnvMissing`.
   */
  ServiceEnvMissing: 'ServiceEnvMissing',

  /**
   * Per-service active healthcheck (`healthcheck.command` from ServiceSpec)
   * returned exit 0 within the stage deadline. Emitted by
   * `StartingServicesStage` on the first successful probe; the inner probe
   * loop then breaks for that service. Payload:
   * `{ serviceName, durationMs, lastExitCode: 0 }`. Mirrors
   * `Source.Features.RuntimeEvents.Models.RuntimeEventTypes.ServiceHealthy`.
   */
  ServiceHealthy: 'ServiceHealthy',

  /**
   * Per-service active healthcheck never returned exit 0 within the stage
   * deadline. Emitted by `StartingServicesStage` — does NOT fail the stage
   * (process-alive is still required, but the healthcheck is advisory once
   * the service is RUNNING). Payload:
   * `{ serviceName, deadlineMs, lastExitCode, lastStdoutTail, lastStderrTail }`.
   * Severity: Warn.
   */
  ServiceHealthcheckTimedOut: 'ServiceHealthcheckTimedOut',

  /**
   * Single healthcheck probe iteration returned non-zero. Bucketed emission
   * (only on exit-code transition OR every 5th attempt) so a 180s × 5s loop
   * doesn't flood the events table. Payload:
   * `{ serviceName, attemptCount, exitCode, stdoutTail, stderrTail,
   * attemptedAt }`. Severity: Info.
   */
  ServiceHealthcheckProbeFailed: 'ServiceHealthcheckProbeFailed',

  /**
   * Batched stdout/stderr lines from a supervised service tailed during the
   * `starting-services` bootstrap window. Same flush cadence as
   * `BootstrapOutputChunk` (~2s or 50 lines). Payload:
   * `{ serviceName, stream, lines, batchedAt, lineCount, flushReason }`.
   * Mirrors `RuntimeEventTypes.ServiceOutputChunk` on the backend.
   */
  ServiceOutputChunk: 'ServiceOutputChunk',

  // -------- Spec delta application --------
  SpecDeltaApplied: 'SpecDeltaApplied',
  SpecDeltaFailed: 'SpecDeltaFailed',
  SpecValidationFailed: 'SpecValidationFailed',

  /**
   * Degraded-online bootstrap (self-healing-runtime-specs, card D1). A
   * NON-CRITICAL (spec) stage — Install / RunningSetup / StartingServices —
   * failed deterministically, so the orchestrator recorded a `BootIssue` and
   * CONTINUED rather than aborting the boot. One event is emitted per boot
   * issue just before `ReportReadyStage` runs, so the runtime still reaches
   * Online but in a `Degraded` SpecHealth state. The amber UI banner + the
   * agent self-heal loop key off these. Payload:
   * `{ stage, service?, reason, detail?, occurredAt }` (mirrors the BootIssue
   * shape). Severity: Warn — the runtime is alive, the spec just didn't fully
   * apply. Mirrors
   * `Source.Features.RuntimeEvents.Models.RuntimeEventTypes.SpecDegraded`.
   */
  SpecDegraded: 'SpecDegraded',

  // -------- Observability foundation (super-admin polish) --------
  // Mirror of the C# constants under `Source.Features.RuntimeEvents.Models.RuntimeEventTypes`
  // added by the backend's observability-foundation card. Keep in lockstep.

  /** Service declared FATAL by supervisord without ever reaching RUNNING. */
  ServiceFailedToStart: 'ServiceFailedToStart',
  /** Daemon's RuntimeSpecDeltaApplied ack rejected by SignalR (payload-too-large, transport blow-up). */
  SpecApplyAckFailed: 'SpecApplyAckFailed',
  /** Backend-emitted: our runtime view disagrees with Fly's live machine view. */
  RuntimeFlyDriftDetected: 'RuntimeFlyDriftDetected',
  /** Cloudflared tunnel hostname health check transitioned Up → Down. */
  CloudflaredTunnelDown: 'CloudflaredTunnelDown',
  /** Cloudflared tunnel hostname health check transitioned Down → Up after a prior Down. */
  CloudflaredTunnelUp: 'CloudflaredTunnelUp',
  /** Backend-emitted: RespawnRuntimeJob decided to respawn a crashed runtime. */
  RuntimeRespawnTriggered: 'RuntimeRespawnTriggered',
  /** Disk pressure on `/data` transitioned to critical. */
  DiskPressureCritical: 'DiskPressureCritical',
} as const

export type RuntimeEventType =
  (typeof RuntimeEventTypes)[keyof typeof RuntimeEventTypes]
